src/content/en/updates/2017/10/animated-blur.md

```diff
@@ -0,0 +1,187 @@
+filters](https://en.wikipedia.org/wiki/Kernel_(image_processing)) to apply a
+blur. Convolution filters are fairly expensive as for every output pixel a
+number of input pixels have to be considered. The bigger the image or the bigger
+the blur radius, the more costly the effect is.
+
+And that's where the problem lies, we are running a rather expensive GPU
+operation every frame, blowing our frame budget of 16ms and therefore ending up
+well below 60fps.
+
+
+## Down the rabbit hole
+
+So what can we do to make this run smoothly? We can use sleight of hand! Instead
+of animating the actual blur value (the radius of the blur), we pre-compute a
+couple of blurred copies where the blur value increases exponentially and
+cross-fade between them using `opacity`.
+
+The cross-fade is a series of overlapping opacity fade-ins and fade-outs. If we
+have four blur stages for example, we fade out the first stage while fading in
+the second stage at the same time. Once the second stage reaches 100% opacity
+and the first one has reached 0%, we fade out the second stage while fading in
+the third stage. Once that is done, we finally fade out the third stage and fade
+in the fourth and final version. In this scenario, each stage would take ¼ of
+the total desired duration. Visually, this looks very similar to a real,
+animated blur.
+
+In our experiments increasing the blur radius exponentially per stage yielded
+the best visual results. Example: If we have four blur stages we'd apply
+`filter: blur(2^n)` to each stage, i.e. stage 0: 1px, stage 1: 2px, stage 2: 4px
+and stage 3: 8px. If we force each of these blurred copies onto their own layer
+(called "promoting") using `will-change: transform`, changing opacity on these
+elements should be super-duper fast. In theory, this would allow us to
+front-load the expensive work of blurring. Turns out, the logic is flawed. If
+you run [this
+demo](https://googlechromelabs.github.io/ui-element-samples/animated-blur/badblur.html),
+you'll see framerate is still below 60fps — it's actually *worse* than before.
+
+<img src="/web/updates/images/2017/10/animated-blur/badblur.jpg" alt="DevTools
+  showing a trace where the GPU has long periods of busy time.">
+
+A quick look into DevTools reveals that the GPU is still extremely busy and
+stretches each frame to ~90ms. But why? We are not changing the blur value
+anymore, only the opacity, so what's happening? The problem lies, once again, in
+the nature of the blur effect: As explained before, if the element is both
+promoted and blurred, the effect is applied by the GPU. So even though we are
+not animating the blur value anymore, the texture itself is still unblurred and
+needs to be re-blurred every frame by the GPU. The reason for the frame rate
+being even worse than before stems from the fact that compared to the naïve
+implementation the GPU actually has more work than before as most of the time 2
+textures are visible that need to be blurred independently.
```

Comment:
It seems like there's a general principle that you're not saying: don't assume that because you have a GPU that all your animation issues are solved. Maybe you could say something about that in the intro or conclusion?

Reply:
Added a note to the intro